src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test.Common/CSharpTestLspServer.cs

```diff
@@ -0,0 +1,135 @@
+            AdhocWorkspace testWorkspace,
+            ServerCapabilities serverCapabilities)
+        {
+            TestWorkspace = testWorkspace;
+
+            var (clientStream, serverStream) = FullDuplexStream.CreatePair();
+
+            _serverMessageFormatter = CreateJsonMessageFormatter();
+            _serverMessageHandler = new HeaderDelimitedMessageHandler(serverStream, serverStream, _serverMessageFormatter);
+            _serverRpc = new StreamJsonRpc.JsonRpc(_serverMessageHandler)
+            {
+                ExceptionStrategy = ExceptionProcessing.ISerializable,
+            };
+
+            _languageServer = CreateLanguageServer(_serverRpc, testWorkspace, serverCapabilities);
+
+            _clientMessageFormatter = CreateJsonMessageFormatter();
+            _clientMessageHandler = new HeaderDelimitedMessageHandler(clientStream, clientStream, _clientMessageFormatter);
+            _clientRpc = new StreamJsonRpc.JsonRpc(_clientMessageHandler)
+            {
+                ExceptionStrategy = ExceptionProcessing.ISerializable,
+            };
+
+            _clientRpc.StartListening();
+
+            static JsonMessageFormatter CreateJsonMessageFormatter()
+            {
+                var messageFormatter = new JsonMessageFormatter();
+                VSInternalExtensionUtilities.AddVSInternalExtensionConverters(messageFormatter.JsonSerializer);
+                return messageFormatter;
+            }
+
+            static IRazorLanguageServerTarget CreateLanguageServer(
+                StreamJsonRpc.JsonRpc serverRpc,
+                Workspace workspace,
+                ServerCapabilities serverCapabilities)
+            {
+                var capabilitiesProvider = new RazorCapabilitiesProvider(serverCapabilities);
+
+                var exportProvider = TestCompositions.Roslyn.ExportProviderFactory.CreateExportProvider();
+                var registrationService = exportProvider.GetExportedValue<RazorTestWorkspaceRegistrationService>();
+                registrationService.Register(workspace);
+
+                var languageServerFactory = exportProvider.GetExportedValue<IRazorLanguageServerFactoryWrapper>();
+                var languageServer = languageServerFactory.CreateLanguageServer(serverRpc, capabilitiesProvider);
+
+                serverRpc.StartListening();
+                return languageServer;
+            }
+        }
+
+        internal static async Task<CSharpTestLspServer> CreateAsync(
+            AdhocWorkspace testWorkspace,
+            ClientCapabilities clientCapabilities,
+            ServerCapabilities serverCapabilities)
+        {
+            var server = new CSharpTestLspServer(testWorkspace, serverCapabilities);
+
+            await server.ExecuteRequestAsync<InitializeParams, InitializeResult>(Methods.InitializeName, new InitializeParams
+            {
+                Capabilities = clientCapabilities,
+            }, CancellationToken.None);
+
+            return server;
```

Comment:
Before returning should this wait for the `initialized` notification or in practice does it not matter?

Reply:
Hmm, it doesn't seem to make a difference for test results, and Roslyn doesn't seem to wait either in their code. But it might be good to add just to be safe? I might check with you later this week to check how to do this properly